### PR TITLE
Add 2D Probabilistic UNet Training Setup

### PR DESCRIPTION
**Pull request by anwai98**
- PUNet Source Training Workflow

> (TODO: evaluation script will follow)

### CODE REVIEW

File: torch_em/self_training/probabilistic_unet_trainer.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            prior_samples=16,
+            supervised_loss=None,
+            supervised_loss_and_metric=None,
+            **kwargs
+    ):
+        super().__init__(**kwargs)
+        assert supervised_loss, supervised_loss_and_metric is not None
+
+        self.supervised_loss = supervised_loss
+        self.supervised_loss_and_metric = supervised_loss_and_metric
```
Comment (constantinpape) [inline torch_em/self_training/probabilistic_unet_trainer.py:35]: just a minor point, but we can just call it `loss` and `loss_and_metric` here, since this trainer is only for supervised training, so we don't need to make a distinction between the supervised and unsupervised cases.

File: torch_em/self_training/probabilistic_unet_trainer.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    Parameters:
+        clip_posterior [bool] - (default: False)
+        prior_samples [int] - (default: 16)
+        supervised_loss [callable] - (default: None)
+        supervised_loss_and_metric [callable] - (default: None)
+    """
+
+    def __init__(
+            self,
+            clip_posterior=False,
```
Comment (constantinpape) [inline torch_em/self_training/probabilistic_unet_trainer.py:25]: I would suggest to just use a `clipping_value` parameter with default `None`. If set to `None` don't clip, otherwise use the `clipping_value`. (see also comment below).

File: torch_em/self_training/probabilistic_unet_trainer.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            **kwargs
+    ):
+        super().__init__(**kwargs)
+        assert supervised_loss, supervised_loss_and_metric is not None
+
+        self.supervised_loss = supervised_loss
+        self.supervised_loss_and_metric = supervised_loss_and_metric
+
+        self.clip_posterior = clip_posterior
+        self.clipping_value = 1
```
Comment (constantinpape) [inline torch_em/self_training/probabilistic_unet_trainer.py:38]: `self.clipping_value = clipping_value`

File: torch_em/self_training/probabilistic_unet_trainer.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+            with forward_context():
+                # We pass the model, the input and the labels to the supervised loss function, so
+                # that's how the loss is calculated stays flexible, e.g. here to enable ELBO for PUNet.
+                loss = self.supervised_loss(self.model, x, y)
+
+            backprop(loss)
+
+            # To counter the exploding gradients in the posterior net
+            if self.clip_posterior:
```
Comment (constantinpape) [inline torch_em/self_training/probabilistic_unet_trainer.py:76]: `if self.clipping_value is not None`

**Review by constantinpape — COMMENTED** [review]
I left a few minor comments; looks good overall.

File: torch_em/self_training/probabilistic_unet_trainer.py (hunk 1 above)
Comment (anwai98) [reply torch_em/self_training/probabilistic_unet_trainer.py:35]: Yes, this makes sense. Updated it in the traning scheme as well as the loss. Thanks!

File: experiments/probabilistic_domain_adaptation/livecell/punet_source.py [hunk 5]
```diff
@@ -26,16 +26,15 @@ def _train_cell_type(args, cell_type, device=DEVICE):
         model=model,
         train_loader=train_loader,
         val_loader=val_loader,
-        loss=DummyLoss(),
         metric=DummyLoss(),
```
Comment (constantinpape) [inline experiments/probabilistic_domain_adaptation/livecell/punet_source.py:29]: It's also better if you do this already in the trainer. Then it's not necessary to do it on the user side. (see comment below)

File: torch_em/self_training/probabilistic_unet_trainer.py [hunk 6]
```diff
@@ -14,28 +14,26 @@ class ProbabilisticUNetTrainer(torch_em.trainer.DefaultTrainer):
...
             prior_samples=16,
-            supervised_loss=None,
-            supervised_loss_and_metric=None,
+            loss=None,
+            loss_and_metric=None,
             **kwargs
     ):
-        super().__init__(**kwargs)
-        assert supervised_loss, supervised_loss_and_metric is not None
+        super().__init__(loss=loss, **kwargs)
```
Comment (constantinpape) [inline torch_em/self_training/probabilistic_unet_trainer.py:31]: set `metric=DummyMetric()` here, see above.

**Comment by anwai98** [comment]
The PUNet Source Workflow is completed now. Just updated the pipeline for source evaluations. Thanks!

File: experiments/probabilistic_domain_adaptation/livecell/common.py [hunk 7]
```diff
@@ -23,6 +23,7 @@
 from torch_em.util import load_model
 
 CELL_TYPES = ["A172", "BT474", "BV2", "Huh7", "MCF7", "SHSY5Y", "SkBr3", "SKOV3"]
+DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (constantinpape) [inline experiments/probabilistic_domain_adaptation/livecell/common.py:26]: This is not a good idea to declare globally, and I also don't see that you use it anywhere.

**Review by constantinpape — COMMENTED** [review]
One minor comment, otherwise this looks good.
I am looking into the test failure here now: https://github.com/constantinpape/torch-em/pull/124

**Review by constantinpape — APPROVED** [review]
